Clean Django templates to use 4 space indents

adnrs96 opened the pull request:
In this PR we continue the work being done in PR #3658 to clean up the django templates located in 
`/templates/zerver/` to follow 4 space indentation. The improvements to pretty_print tool has been pushed as a commit to PR #3658. All the files cleaned up in this PR does not pose any merge conflicts with other open PR 's.

smarx commented:
*Automated message from Dropbox CLA bot*

@adnrs96, it looks like you've already signed the Dropbox CLA. Thanks!

showell commented:
@timabbott I think this is ready to merge, although you might want to look at markdown_help.html as interesting case of where the pretty printer leaves some of the original dubious formatting intact.

adnrs96 commented:
I have updated the PR with changes to `markdown_help.html` and some more.

timabbott commented:
I merged a few commits, thanks @adnrs96!  Posted some comments on chat.zulip.org on a few issues; I think there's a couple small bugs in the formatter for Jinja2 templates.

timabbott commented on templates/zerver/base.html:
```diff
@@ -3,42 +3,42 @@
+        </body>
 
-</html>
+    </html>
```
This looks wrong

adnrs96 replied:
yeah, had a look and turns out to be issue with non block django tags.

timabbott commented on templates/zerver/invite_user.html:
```diff
@@ -1,39 +1,39 @@
+                <div class="controls">
+                    <textarea rows="2" id="invitee_emails"
+                        name="invitee_emails"
+                              placeholder="{{ _('One or more email addresses...') }}"></textarea>
```
`placeholder` is over-indented here.

timabbott commented on templates/zerver/markdown_help.html:
````diff
@@ -1,105 +1,105 @@
-          <td class="preserve_spaces">``` .py
+                    </tr>
+                    <tr>
+                      <td class="preserve_spaces">``` .py
````
This `<td>` spacing seems off?

adnrs96 replied:
That is a manual error. markdown_help actually needs manual editing after running tool.
Will fix it.

adnrs96 commented:
Thanks for reviewing and merging. I have updated this PR with new commits. It turns out there were some overlooked issue in tooling with respect to django tags which do not have a block structure and end tags indents. Now that is fixed and the PR #3658 has also been updated with commit containing changes in tooling.

timabbott commented:
I merged another batch of these, with a few manual changes.
* `api.html` I can't merge because it has some meaningful whitespace in the code blocks -- e.g. the `zuliprc` output.  I kinda want to just redo those sections in markdown...
* features/apps/integrations/hello I can't merge because there's some active work on them that doesn't have a PR open yet, but will soon.

adnrs96 commented:
Thank you for this. I am curious about the whitespace and zuliprc stuff you mentioned. I quite don't follow what you meant by zuliprc output?
For the features/apps/integrations/hello I am thinking of having them as well in a exception list when we add automatic checking for 4 space indents.

adnrs96 commented:
Just added new commits 3 of which have been merged already last time, namely files 
* `portico.html `
* `accounts_accept_terms.html`
* `api_endpoints.html`

I am afraid to tell you that we missed out some thing in these files which the tool didn't fixed indentation properly for (was a kind of bug in tool, now resolved and updated in commit [here](https://github.com/zulip/zulip/pull/3658/commits/1f0517b5d41858494da982ca3f194028b93cdf29) under this [PR#3658](https://github.com/zulip/zulip/pull/3658)). The new commits have those updates.

Please review and advice.

timabbott commented:
So the features.html and integrations.html changes that Brock was working on are now merged, so you can add commits for those when you get a chance.

timabbott commented:
left sidebar has some merge conflicts so you should probably regenerate that one; I merged the rest.

timabbott commented:
Probably now's a good time to analyze the files that still need to migrate and which PRs are important to merge for this...

adnrs96 commented:
Sure let me run the analyser again.

adnrs96 commented:
Just updated the PR with new commits. Here is a list of PR's which should be cleared to proceed further.

templates/zerver/compose.html [PR#3118(+1-2)](https://github.com/zulip/zulip/pull/3118) [PR#2721(+1-0)](https://github.com/zulip/zulip/pull/2721)
templates/zerver/login.html [PR#1747(+7-2)](https://github.com/zulip/zulip/pull/1747) [PR#3905(+1-1)](https://github.com/zulip/zulip/pull/3905)
templates/zerver/navbar.html [PR#2379(+1-0)](https://github.com/zulip/zulip/pull/2379) [PR#3420(+2-2)](https://github.com/zulip/zulip/pull/3420) [PR#2973(+69-60)](https://github.com/zulip/zulip/pull/2973)
templates/zerver/right_sidebar.html [PR#3420(+2-2)](https://github.com/zulip/zulip/pull/3420)
templates/zerver/features.html [PR#3905(+1-1)](https://github.com/zulip/zulip/pull/3905)
templates/zerver/register.html [PR#3905(+2-2)](https://github.com/zulip/zulip/pull/3905) [PR#3007(+3-0)](https://github.com/zulip/zulip/pull/3007)
templates/zerver/keyboard_shortcuts.html [PR#2292(+8-0)](https://github.com/zulip/zulip/pull/2292) [PR#2854(+4-0)](https://github.com/zulip/zulip/pull/2854)
templates/zerver/index.html [PR#2292(+1-0)](https://github.com/zulip/zulip/pull/2292)
templates/zerver/home.html [PR#2089(+2-0)](https://github.com/zulip/zulip/pull/2089) [PR#3420(+1-1)](https://github.com/zulip/zulip/pull/3420) [PR#3002(+2-0)](https://github.com/zulip/zulip/pull/3002) [PR#749(+5-0)](https://github.com/zulip/zulip/pull/749)
templates/zerver/subscriptions.html [PR#1954(+1-0)](https://github.com/zulip/zulip/pull/1954)

Also i wanted to know about api.html and markdown stuff relating to it you talked about.

timabbott commented:
#3905 is merged, and I think we shouldn't worry about ultra-old PRs (e.g. <2000) with 1 line of changes.

adnrs96 commented:
Updated with new commit's. Also cleaned files which are in PR's 1954, 749, 1747. These PR's either had merge conflicts for these files already or very minimal change. Updated the list as well.

timabbott commented:
I merged a few.  With `integrations.html`, there's a problem with leading whitespace in the `<pre>` code blocks (check out what the "basecamp" one looks like)

adnrs96 commented:
Yes just noticed it. This `pre` tag has given me some hard time. For the timing i have adjusted the pretty printer to just leave alone the `pre` tag code wherever it is in the html. Obviously this may lead to ugly looking code but we can't do much of anything until we deploy some kind of JS (just looked around on google) to deal with `pre` tags and the indents. But this might change the meaning of `pre` tag itself i believe.

timabbott commented:
I think the medium-term plan is to port a lot of that documentation into markdown, so maybe not worth worrying about.

adnrs96 commented:
@timabbott you may want to look at this, i have refreshed old one's and done some new one's.
You might find some weird things in here like with the `<strong>` tag in `zephyr-mirror`. To fix those we might want to add exception for such tags in validate function in template_parser.py
There are lot more templates to go here, will keep updating it with stuff.
(All the above are not going to cause merge conflicts with other PR's currently).

brockwhittaker commented:
`landing-nav` and `left-sidebar` I think both have somewhat active stuff?

timabbott commented:
@adnrs96 the `zephyr.html` stuff is problematic because of how the code-blocks are line-wrapped.  

I think maybe we should just convert some of those pages to markdown?

adnrs96 commented:
@timabbott Can you point me towards something that can help convert pages to markdown. Like has it been done before at some point in project which I can refer to. It looks like there are many files which could use the markdown approach like the files inside `templates/zerver/integrations/` which needs to be cleaned up. One thing I had in mind was whether we can add colours to stuff in markdown as we did with the code blocks inside the integrations pages using codehilite.

timabbott commented:
@eeshangarg is working on this; see #5147.

timabbott commented:
I think that's definitely a project that others could collaborate on.

adnrs96 commented:
@timabbott I have updated this PR with more commits and some markdown migrations. All these changes are safe to merge and will not cause merge conflicts.
One thing I wanted to ask was about what to do to migrate these three files to markdown.
```
'templates/corporate/zephyr-mirror.html',
'templates/zerver/api.html',
'templates/zerver/markdown_help.html',
```
Because these are not handled the way webhooks or integrations are, I believe some kind of modification to code base will be required. Primarily I believe it would mean to somehow pass our Jinja temps first through markdown render and then to the Jinja template render. Currently I don't have much of idea about how to do that if we want to even do that. It would be great if you could point me in direction towards making such change.

Apart from this we seems to be nearing towards completing this migration, only a handful files which are blocked on PR's are left. Here is list of such PR's:
'templates/analytics/stats.html [PR#4299(+3-0)](https://github.com/zulip/zulip/pull/4299) [PR#4325(+11-7)](https://github.com/zulip/zulip/pull/4325)',
'templates/analytics/activity.html [PR#5059(+1-1)](https://github.com/zulip/zulip/pull/5059)',
'templates/zerver/navbar.html [PR#5201(+1-1)](https://github.com/zulip/zulip/pull/5201) [PR#4083(+140-132)](https://github.com/zulip/zulip/pull/4083)',
'templates/zerver/hello.html [PR#5059(+1-1)](https://github.com/zulip/zulip/pull/5059)',
'templates/zerver/home.html [PR#4177(+3-0)](https://github.com/zulip/zulip/pull/4177) [PR#749(+5-0)](https://github.com/zulip/zulip/pull/749)',
'templates/zerver/login.html [PR#1747(+7-2)](https://github.com/zulip/zulip/pull/1747) [PR#4944(+2-2)](https://github.com/zulip/zulip/pull/4944)',
'templates/zerver/compose.html [PR#2721(+1-0)](https://github.com/zulip/zulip/pull/2721) [PR#4986(+1-3)](https://github.com/zulip/zulip/pull/4986)',
'templates/zerver/keyboard_shortcuts.html [PR#4931(+4-0)](https://github.com/zulip/zulip/pull/4931)',
'templates/zerver/index.html [PR#5059(+1-1)](https://github.com/zulip/zulip/pull/5059)',
'templates/zerver/register.html [PR#4944(+2-2)](https://github.com/zulip/zulip/pull/4944) [PR#5059(+1-1)](https://github.com/zulip/zulip/pull/5059)',
'templates/zerver/apps.html [PR#5059(+1-1)](https://github.com/zulip/zulip/pull/5059)',
'templates/zerver/integrations/index.html [PR#5059(+1-1)](https://github.com/zulip/zulip/pull/5059)',
'templates/zerver/emails/digest.html [PR#4699(+2-2)](https://github.com/zulip/zulip/pull/4699)',
'static/html/5xx.html [PR#5112(+19-17)](https://github.com/zulip/zulip/pull/5112)',
'templates/zerver/features.html [PR#4342(+25-0)](https://github.com/zulip/zulip/pull/4342) [PR#5059(+1-1)](https://github.com/zulip/zulip/pull/5059)',
'templates/zerver/search_operators.html [PR#4279(+4-0)](https://github.com/zulip/zulip/pull/4279)',
'templates/zerver/api_endpoints.html [PR#5070(+1-2)](https://github.com/zulip/zulip/pull/5070)',

@eeshangarg you might want to take a look at those migrations. All the commits with Migration in front refer to such migrations.

eeshangarg commented on templates/zerver/integrations/capistrano.md:
```diff
@@ -0,0 +1,34 @@
+First, download and install our [Python bindings and example scripts](/api) on your Capistrano server. Once you've
```
I think this line can be made part of a macro. Asking users to download our bindings is something a lot of these integrations do. :)

eeshangarg commented on templates/zerver/integrations/capistrano.md:
```diff
@@ -0,0 +1,34 @@
+notification to fire at a different time.
+
+**Congratulations! You're done!**  
+Whenever you do a deployment, you'll get an automated notification that looks like this:
```
Consider replacing line 29-30 with the macro `templates/zerver/help/include/congrats.md`. :)

eeshangarg commented on templates/zerver/integrations/git.md:
```diff
@@ -0,0 +1,34 @@
+First, download and install our [Python bindings and example scripts](/api) on your Git server.
```
Again, this part can be replaced by a macro. You'll have to create one under `templates/zerver/help/include/*`. :)

eeshangarg commented on templates/zerver/integrations/git.md:
```diff
@@ -0,0 +1,34 @@
+First, download and install our [Python bindings and example scripts](/api) on your Git server.
+
+Next, create the stream you’d like to use for Git notifications, and subscribe all interested parties to this
+stream. The integration will use the default stream `commits` if no stream is supplied in the hook; you still need
```
Replace the create the stream part with the `create-stream.md` macro.

eeshangarg commented on templates/zerver/integrations/git.md:
```diff
@@ -0,0 +1,34 @@
+
+Whenever you make a push to the `master` branch of your git repository (or whatever you configured above), the Zulip
+git plugin will send an automated notification that looks like this:
+
```
Replace with `congrats.md`. :) Consider moving line 27-28 to after the `Symlink...` line and then use `congrats.md`. :)

eeshangarg commented on templates/zerver/integrations/google-calendar.md:
```diff
@@ -0,0 +1,55 @@
+Get Google Calendar reminders in Zulip! This is a great way to see your reminders directly in your Zulip feed.
+
+First download and install our [Python Bindings and example scripts](/api). This bot should be set up on a trusted
```
Same recommendation as before. :)

eeshangarg commented on templates/zerver/integrations/google-calendar.md:
```diff
@@ -0,0 +1,55 @@
+
+**Congratulations! You’re done!**
+
+You will get a Zulip private message, whenever you have a calendar event scheduled, that looks like this:
```
Use the `congrats.md` macro please.

eeshangarg commented on templates/zerver/integrations/jira-plugin.md:
```diff
@@ -0,0 +1,50 @@
+
+First, create the stream you'd like to use for JIRA notifications, and subscribe all interested parties to this
+stream. We recommend the name `jira`.
+
```
Use `create-stream.md` macro.

adnrs96 replied:
I guess since this is the only place we could have used `create-stream.md` because of the way it is phrased we should leave it out here as well.

eeshangarg commented on templates/zerver/integrations/jira-plugin.md:
```diff
@@ -0,0 +1,50 @@
+**Congratulations! You're done!**
+
+Updates from JIRA will be sent to the stream you've configured, as below.
+
```
Here again, you can move line 48 to after line 44 and then use `congrats.md`. :)

eeshangarg commented on templates/zerver/integrations/mercurial.md:
```diff
@@ -0,0 +1,74 @@
+    stream = "commits"
+    web_url = "http://hg.example.com:8000/"
+    {% if api_site_required %}site = {{ external_api_uri_subdomain }}{% endif %}
+
```
Shouldn't this code block here (lines 40-45) and other code blocks be wrapped in `~~~ <code> ~~~` blocks?

adnrs96 replied:
It renders it as code block if indented. Indentation was needed to keep in line with previous version of doc.

eeshangarg commented on templates/zerver/integrations/nagios.md:
```diff
@@ -0,0 +1,50 @@
+still need to create the stream even if you are using this default.
+
+Next, download and install our [Python bindings and example scripts](/api) on your Nagios server.
+
```
I think my recommendations to use `create-stream.md`, `congrats.md` and to create and use a macro for the instruction to download and install our bindings apply everywhere. :)

eeshangarg commented on templates/zerver/integrations/openshift.md:
```diff
@@ -0,0 +1,54 @@
+Ths integration sends a notification every time a deployment is made in an OpenShift instance.
+
+First, download our [Pythonbindings and example scripts](/api), and move it to your local copy of the application’s
```
A little error here, it should be `[Python bindings`. Another reason to use a macro for this. :)

eeshangarg commented on templates/zerver/integrations/openshift.md:
````diff
@@ -0,0 +1,54 @@
+ZULIP_API_KEY = "0123456789abcdef0123456789abcdef"
+{% if api_site_required %}ZULIP_SITE = "{{ external_api_uri_subdomain }}"{% endif %}
+```
+
````
I feel like someone mentioned this on the Dropbox Paper instance documenting our migration efforts. We might want to create a macro for lines 23-30 (or just lines 27-29). It seems like we are using similar code blocks in a lot of these files. :)

eeshangarg commented on templates/zerver/integrations/subversion.md:
```diff
@@ -0,0 +1,30 @@
+        ZULIP_USER = "svn-bot@example.com"
+        ZULIP_API_KEY = "0123456789abcdef0123456789abcdef"
+        {% if api_site_required %}ZULIP_SITE = "{{ external_api_uri_subdomain }}"{% endif %}
+
```
Yes we definitely want a macro for this block. :)

adnrs96 replied:
Maybe not a good idea since values of ZUILP_USER are different everywhere and are somewhat depicting type of bot.

adnrs96 commented:
I have updated the commits with most of the changes @eeshangarg suggested. Left out the use of create stream macro as it might not be that suitable to use due to variety of wording used here. Eg. The suggestions for name are not necessarily integration name. Also I added a commit 379b30b which deals with adding stuff like integration names to the context of the templateview to be rendered.
I guess this should be ready for a merge now. @timabbott

timabbott commented:
I went through these and merged the things that I could.  Thanks for doing this @adnrs96 and reviewing this @eeshangarg!  I made a few changes:
* I line-wrapped all of your new markdown docs; super long lines should be avoided wherever possible.
* I didn't merge the email changes, since they conflict with #5258, which will also fix this problem for most of those anyway.

I agree with @eeshangarg that we should clean up the `templatetags` code soon after this to be more readable.

Since this PR is getting cluttered, I'm going to close it and we can continue on a new one.

Thanks for all your work on this so far @adnrs96 !
